File: backend/app/schemas/organization.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class OrganizationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: str
    name_ar: str
    name_en: str
    org_type: str
    deployment_mode: str
    country_code: str
    timezone: str
    default_language: str
    default_currency: str
    logo_url: str | None
    primary_color: str
    status: str
    subscription_plan: str | None
    subscription_expires_at: datetime | None
    report_cadence_days: int | None = Field(default=None, ge=1, le=366)
    created_at: datetime
    updated_at: datetime


class OrganizationUpdate(BaseModel):
    """Settings an org admin may edit. Code, org_type, and deployment_mode
    are intentionally NOT here — those are platform-level and should not
    drift via the admin UI."""

    name_ar: str | None = Field(default=None, min_length=1, max_length=255)
    name_en: str | None = Field(default=None, min_length=1, max_length=255)
    country_code: str | None = Field(default=None, min_length=2, max_length=2)
    timezone: str | None = Field(default=None, max_length=50)
    default_language: str | None = Field(default=None, min_length=2, max_length=5)
    default_currency: str | None = Field(default=None, min_length=3, max_length=3)
    logo_url: str | None = Field(default=None, max_length=500)
    primary_color: str | None = Field(default=None, pattern=r"^#[0-9A-Fa-f]{6}$")
    # None (or omitted) means "use the platform default cadence".
    report_cadence_days: int | None = Field(default=None, ge=1, le=366)

File: frontend/src/lib/organization.ts
import { api } from "./api";

export interface Organization {
  id: string;
  code: string;
  name_ar: string;
  name_en: string;
  org_type: string;
  deployment_mode: string;
  country_code: string;
  timezone: string;
  default_language: string;
  default_currency: string;
  logo_url: string | null;
  primary_color: string;
  status: string;
  subscription_plan: string | null;
  subscription_expires_at: string | null;
  report_cadence_days: number | null;
  created_at: string;
  updated_at: string;
}

export interface OrganizationUpdateInput {
  name_ar?: string;
  name_en?: string;
  country_code?: string;
  timezone?: string;
  default_language?: string;
  default_currency?: string;
  logo_url?: string;
  primary_color?: string;
  /** null clears the cadence back to the platform default (90 days). */
  report_cadence_days?: number | null;
}

export async function fetchOrganization(): Promise<Organization> {
  const { data } = await api.get<Organization>("/organization");
  return data;
}

export async function updateOrganization(
  payload: OrganizationUpdateInput,
): Promise<Organization> {
  const { data } = await api.patch<Organization>("/organization", payload);
  return data;
}
